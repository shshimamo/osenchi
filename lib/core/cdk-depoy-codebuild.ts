import * as cdk from 'aws-cdk-lib';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as iam from 'aws-cdk-lib/aws-iam';

export class CdkDeployCodeBuildStack extends cdk.Stack {
    constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const buildImage = codebuild.LinuxBuildImage.fromDockerRegistry(
            "public.ecr.aws/bitnami/node:16.18.1"
        )
        const project = new codebuild.Project(this, "OsenchiProject", {
            projectName: "osenchi-cdk-deploy",
            environment: {
                buildImage
            },
            buildSpec: codebuild.BuildSpec.fromObject({
                version: "0.2",
                phases: {
                    install: {
                        commands: [
                            'npm install -g aws-cdk',
                            "npm install",
                        ]
                    },
                    pre_build: {
                       commands: [
                           'cdk --version',
                           'cdk diff',
                       ]
                    },
                    build: {
                        commands: [
                            'npx cdk deploy --verbose --all --require-approval never'
                        ]
                    }
                }
            })
        })

        project.role?.addManagedPolicy(
            iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess")
        )
    }
}