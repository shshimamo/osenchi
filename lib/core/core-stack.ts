import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as cloudtrail from 'aws-cdk-lib/aws-cloudtrail';
import * as events from 'aws-cdk-lib/aws-events';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';

export interface IBucketProps {
    inputName: string;
    outputName: string;
    logName: string;
}

export interface ITopicProps {
    topicName: string;
    emails: string[];
}

export interface ICoreProps extends cdk.StackProps {
    bucket: IBucketProps;
    topic: ITopicProps;
    detectionFuncName: string;
    deletionFuncName: string;
}

export class CoreStack extends cdk.Stack {
    /** 入力バケット */
    readonly inputBucket: s3.Bucket;

    /** 出力バケット */
    readonly outputBucket: s3.Bucket;

    /** メールトピック */
    readonly emailTopic: sns.Topic;

    /** 感情検出Lambda関数 */
    readonly detectionFunc: lambda.Function;

    /** オブジェクト削除Lambda関数 */
    readonly deletionFunc: lambda.Function;

    /** ステートマシン */
    readonly stateMachine: sfn.StateMachine;

    /**
     * スタックのインスタンスを生成します。
     *
     * @param {Construct} scope
     * @param {string} id
     * @param {ICoreProps} [props]
     * @memberof CoreStack
     */
    constructor(scope: Construct, id: string, props?: ICoreProps) {
        super(scope, id, props);

        if (props) {
            // バケット
            [this.inputBucket, this.outputBucket] = this.newBuckets(props.bucket);

            // トピック
            this.emailTopic = this.newEmailTopic(props.topic);

            // Lambda関数
            this.detectionFunc = this.newDetectSentiment(props.detectionFuncName);
            this.deletionFunc = this.newDeleteObject(props.deletionFuncName);

            // ワークフロー
            this.stateMachine = this.newStateMachine();

            // イベントルール
            this.newEventRule();
        }
    }

    /**
     * バケットの生成
     *
     * @private
     * @param {IBucketProps} bucketProps バケット情報
     * @returns {[s3.Bucket, s3.Bucket]}
     * @memberof CoreStack
     */
    private newBuckets(bucketProps: IBucketProps): [s3.Bucket, s3.Bucket] {
        const inputBucket = new s3.Bucket(this, 'InputBucket', {
            bucketName: bucketProps.inputName,
        });

        const outputBucket = new s3.Bucket(this, 'OutputBucket', {
            bucketName: bucketProps.outputName,
        });

        const logBucket = new s3.Bucket(this, 'LogBucket', {
            bucketName: bucketProps.logName,
        });

        const trail = new cloudtrail.Trail(this, 'Trail', {
            bucket: logBucket,
            isMultiRegionTrail: false,
        });
        const s3EventSelector: cloudtrail.S3EventSelector = {
            bucket: inputBucket,
        };
        trail.addS3EventSelector([s3EventSelector], {
            readWriteType: cloudtrail.ReadWriteType.WRITE_ONLY,
        });

        return [inputBucket, outputBucket];
    }

    /**
     * メールトピックの生成
     *
     * @private
     * @param {ITopicProps} topicProps トピック情報
     * @returns {sns.Topic}
     * @memberof CoreStack
     */
    private newEmailTopic(topicProps: ITopicProps): sns.Topic {
        const topic = new sns.Topic(this, 'Topic', {
            topicName: topicProps.topicName,
        });
        for (const email of topicProps.emails) {
            this.validateEmail(email);
            topic.addSubscription(new subscriptions.EmailSubscription(email));
        }
        return topic;
    }

    /**
     * メールアドレスを検証
     *
     * @param {string} email メールアドレス
     */
    private validateEmail(email: string): void {
        const regexp = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

        if (!email.match(regexp)) {
            throw new Error('Email parameter is invalid.');
        }
    }

    /**
     * 感情検出Lambda関数の生成
     *
     * @private
     * @param {string} functionName 関数名
     * @returns {lambda.Function}
     * @memberof CoreStack
     */
    private newDetectSentiment(functionName: string): lambda.Function {
        const func = new lambda.Function(this, 'DetectionFunc', {
            functionName: functionName,
            code: lambda.Code.fromAsset('functions/detect-sentiment', {
                exclude: ['*.ts'],
            }),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_12_X,
            timeout: cdk.Duration.minutes(10),
            environment: {
                DEST_BUCKET: this.outputBucket.bucketName,
            },
        });
        this.inputBucket.grantRead(func);
        this.outputBucket.grantWrite(func);

        const policy = new iam.PolicyStatement({
            resources: ['*'],
            actions: ['comprehend:BatchDetectSentiment'],
        });
        func.addToRolePolicy(policy);

        return func;
    }

    /**
     * オブジェクト削除Lambda関数の生成
     *
     * @private
     * @param {string} functionName 関数名
     * @returns {lambda.Function}
     * @memberof CoreStack
     */
    private newDeleteObject(functionName: string): lambda.Function {
        const func = new lambda.Function(this, 'DeletionFunc', {
            functionName: functionName,
            code: lambda.Code.fromAsset('functions/delete-object', {
                exclude: ['*.ts'],
            }),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_12_X,
        });
        this.inputBucket.grantDelete(func);

        return func;
    }

    /**
     * ステートマシンの生成
     *
     * @private
     * @returns {sfn.StateMachine}
     * @memberof CoreStack
     */
    private newStateMachine(): sfn.StateMachine {
        const sentimentTask = new tasks.LambdaInvoke(this, 'DetectSentiment', {
          lambdaFunction: this.detectionFunc,
        });
        const deleteTask = new tasks.LambdaInvoke(this, 'DeleteObject', {
          lambdaFunction: this.deletionFunc,
        });
        const successTask = new tasks.SnsPublish(this, 'SendSuccessMail', {
            topic: this.emailTopic,
            message: sfn.TaskInput.fromJsonPathAt('$.*'),
            subject: 'Osenchi Success',
        });
        const errorTask = new tasks.SnsPublish(this, 'SendErrorMail', {
            topic: this.emailTopic,
            message: sfn.TaskInput.fromJsonPathAt('$.*'),
            subject: 'Osenchi Error',
        });

        const fail = new sfn.Fail(this, 'Fail');
        errorTask.next(fail);

        const mainFlow = sentimentTask.next(deleteTask).next(successTask);

        const parallel = new sfn.Parallel(this, 'Parallel');
        parallel.branch(mainFlow);
        parallel.addCatch(errorTask, { resultPath: '$.error' });

        const stateMachine = new sfn.StateMachine(this, 'CoreStateMachine', {
            definition: parallel,
            timeout: cdk.Duration.minutes(30),
        });

        this.emailTopic.grantPublish(stateMachine.role);

        return stateMachine;
    }

    /**
     * イベントルールの生成
     *
     * @private
     * @returns {events.Rule}
     * @memberof CoreStack
     */
    private newEventRule(): events.Rule {
        const rule = new events.Rule(this, 'EventRule', {
            eventPattern: {
                source: ['aws.s3'],
                detailType: ['AWS API Call via CloudTrail'],
                detail: {
                    eventSource: ['s3.amazonaws.com'],
                    eventName: ['PutObject'],
                    requestParameters: {
                        bucketName: [this.inputBucket.bucketName],
                    },
                },
            },
        });

        const target = new targets.SfnStateMachine(this.stateMachine);
        rule.addTarget(target);

        return rule;
    }
}