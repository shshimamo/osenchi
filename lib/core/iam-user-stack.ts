import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';

export class IamUserStack extends cdk.Stack {
    constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        const accountId = cdk.Stack.of(this).account;
        const region = cdk.Stack.of(this).region;

        const codebuildRunPolicy = new iam.Policy(this, "codebuild-run", {
            policyName: "codebuild-run"
        })
        codebuildRunPolicy.addStatements(
            new iam.PolicyStatement({
                actions: ["codebuild:StartBuild", "codebuild:BatchGetBuilds"],
                effect: iam.Effect.ALLOW,
                resources: [`arn:aws:codebuild:${region}:${accountId}:project/*`]
            })
        )
        codebuildRunPolicy.addStatements(
            new iam.PolicyStatement({
                actions: ["logs:GetLogEvents"],
                effect: iam.Effect.ALLOW,
                resources: [
                    `arn:aws:logs:${region}:${accountId}:log-group:/aws/codebuild/*:*`
                ]
            })
        )
        const user = new iam.User(this, "githubactions", {
            userName: "githubactions",
        })
        user.attachInlinePolicy(codebuildRunPolicy)
    }
}