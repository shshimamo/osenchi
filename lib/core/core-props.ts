import { Node } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import { Context } from '../common/context';
import { ICoreProps } from './core-stack';

export class CoreProps {
    public static fromContext(node: Node): ICoreProps {
        const context = new Context(node);

        return {
            bucket: {
                inputName: context.service('input'),
                outputName: context.service('output'),
                logName: context.service('log'),
            },
            topic: {
                topicName: context.service('topic'),
                emails: context.get<string>('emails').split(','),
            },
            detectionFuncName: context.service('detect-sentiment'),
            deletionFuncName: context.service('delete-object'),
        };
    }
}