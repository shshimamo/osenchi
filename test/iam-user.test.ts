import * as cdk from 'aws-cdk-lib';
import { Template } from "aws-cdk-lib/assertions";
import { Context } from '../lib/common/context';
import { IamUserStack } from '../lib/core/iam-user-stack';

beforeAll(() => {
    Context.setEnvironment();
});

describe('iam-user Snapshot tests', () => {
    test('Snapshot tests', () => {
        const app = new cdk.App();
        const stack = new IamUserStack(app, 'Osenchi-Core');
        // スタックからテンプレート(JSON)を生成
        const template = Template.fromStack(stack).toJSON();
        expect(template).toMatchSnapshot();
    });
});
