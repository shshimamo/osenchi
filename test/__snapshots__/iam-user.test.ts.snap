// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iam-user Snapshot tests Snapshot tests 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "codebuildrun39D7E2A1": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":project/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "logs:GetLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:/aws/codebuild/*:*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "codebuild-run",
        "Users": Array [
          Object {
            "Ref": "githubactions1B4FC73C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "githubactions1B4FC73C": Object {
      "Properties": Object {
        "UserName": "githubactions",
      },
      "Type": "AWS::IAM::User",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
